### ISSUE TITLE
Telegraf SNMP Error for Multiple Cisco MIBs

### ISSUE THREAD

**Issue by js-mode** [influxdata/telegraf#5647]
Using Telegraf snmp plugin to send metrics to wavefront-proxy with SNMPv3. Got it working for the Interface Stats, but can't get the SNMPv2-SMI::enterprises to work properly. 

GOAL: Get these 5 OIDs configured in Telegraf so I can send these metrics to wavefront: 
```
PowerSupplyStatus: 1.3.6.1.4.1.9.9.117.1.1.2
PowerSupplyValues: 1.3.6.1.4.1.9.9.117.1.1.4
cseSysCPUUtilization 1.3.6.1.4.1.9.9.305.1.1.1
cseSysUpTime 1.3.6.1.4.1.9.9.305.1.1.10
cseSysMemoryUtilization 1.3.6.1.4.1.9.9.305.1.1.2
```
 Running this command gives me the following output: 
`snmptranslate -Td -Ob -m all  1.3.6.1.4.1.9.9.305.1.1.1     1.3.6.1.4.1.9.9.117.1.1.2`

> SNMPv2-SMI::enterprises.9.9.305.1.1.1
> enterprises OBJECT-TYPE
>   -- FROM	SNMPv2-SMI, RFC1155-SMI
> ::= { iso(1) org(3) dod(6) internet(1) private(4) enterprises(1) 9 9 305 1 1 1 }
> 
> SNMPv2-SMI::enterprises.9.9.117.1.1.2
> enterprises OBJECT-TYPE
>   -- FROM	SNMPv2-SMI, RFC1155-SMI
> ::= { iso(1) org(3) dod(6) internet(1) private(4) enterprises(1) 9 9 117 1 1 2 }

If I run a single snmpget command against these OIDs, I get the following error: 

`snmpget  -v2c -c simple _IP_ 1.3.6.1.4.1.9.9.305.1.1.2`
> SNMPv2-SMI::enterprises.9.9.305.1.1.2 = No Such Instance currently exists at this OID

**(note: I only turned on snmpv2c to see if authentication was an issue, but this happened with v2c & v3)**

Attached is my config file (everything from line 60 down works fine), but the section with the above OIDs will not work. 

The instructions are not clear on how to get these configured properly.  I can receive some response when I run snmpwalk on these OIDs, but not with snmpget. 

After creating a new config file with just these OIDs and it gave us this error:

> 2019-03-28T23:30:00Z E! [inputs.snmp]: Error in plugin: initializing table powerStatus: initializing field ucdDemoMIB: translating: exit status 2: SNMPv2-SMI::ucdDemoMIB: Unknown Object Identifier

If I run the snmptable command against the Cisco Nexus9000 switch, this is what I get: 
`snmptable -Ch -Cl -c simple _IP_ SNMPv2-SMI::enterprises`
>ucdDemoMIB ucdInternal ucdExperimental ucdSnmpAgent memory version snmperrs systemStats ucdTraps logMatch
SNMPv2-SMI::enterprises: No entries

snmptable -Ch -Cl -c simple _IP_ 1.3.6.1.4.1.9.9.305.1.1
>Was that a table? SNMPv2-SMI::enterprises.9.9.305.1.1

snmpget  -v2c -c simple _IP_ 1.3.6.1.4.1.9.9.305.1.1.2
>SNMPv2-SMI::enterprises.9.9.305.1.1.2 = No Such Instance currently exists at this OID

snmpwalk  -v2c -c simple _IP_ 1.3.6.1.4.1.9.9.305
SNMPv2-SMI::enterprises.9.9.305.1.1.1.0 = Gauge32: 3
SNMPv2-SMI::enterprises.9.9.305.1.1.2.0 = Gauge32: 59
SNMPv2-SMI::enterprises.9.9.305.1.1.3.0 = Hex-STRING: 07 E3 03 1C 16 33 13 07 00 00 00
SNMPv2-SMI::enterprises.9.9.305.1.1.6.0 = INTEGER: 1
SNMPv2-SMI::enterprises.9.9.305.1.1.7.0 = INTEGER: 2
SNMPv2-SMI::enterprises.9.9.305.1.1.8.0 = INTEGER: 2
SNMPv2-SMI::enterprises.9.9.305.1.1.9.0 = INTEGER: 2
SNMPv2-SMI::enterprises.9.9.305.1.1.10.0 = Gauge32: 13615
SNMPv2-SMI::enterprises.9.9.305.1.5.1.0 = INTEGER: 2
SNMPv2-SMI::enterprises.9.9.305.1.6.1.0 = INTEGER: 2
SNMPv2-SMI::enterprises.9.9.305.1.6.2.0 = INTEGER: 2
SNMPv2-SMI::enterprises.9.9.305.1.7.1.0 = Gauge32: 1
SNMPv2-SMI::enterprises.9.9.305.1.8.1.0 = INTEGER: 2
SNMPv2-SMI::enterprises.9.9.305.1.8.2.0 = INTEGER: 2

I think a deeper issue is understanding how the MIBs versus the OIDs work exactly? Is there a numeric value that signifies the table versus the field? I read on one of the issues that [[inputs.snmp.table]] runs the snmpwalk command.  If that is the case, why can I run the OID in snmpwalk and get a response, but get an error when I put that OID in [[inputs.snmp.table]]?  It is probably something super obvious that I am missing, but would like to be able to customize the different metrics I want to monitor, regardless of what MIB it is apart of. 

Also there was a mention in the documentation that states you need to install the MIBs.  What does this refer to exactly? Is this installing it on the actually switch? The telegraf agent? I am running Telegraf within a docker container, so is there an additional build command I have to pass to it for installing MIBs? 

I have tried looking up as much documentation and seeing if I could get it working, but hopefully other people that get the `No Such Instance currently exists at this OID` error or have similar issues can benefit from me trying to get this sorted out. 

@danielnelson any advice on how to troubleshoot? 

[example.telegraf.txt](https://github.com/influxdata/telegraf/files/3020559/example.telegraf.txt)

**Comment by js-mode** [comment]
For those that don't want to download the file, here is the config for reference: 
```
[global_tags]
  env = "dev"
[agent]
  interval = "60s"
  hostname = "telegraf-snmp"
  omit_hostname = false
[[outputs.file]]
  data_format = "json"
[[outputs.wavefront]]
  host = "WF_PROXY"
  port = WF_PORT
  metric_separator = "."
  convert_paths = true
  use_regex = false

[[inputs.snmp]]
  agents = [AGENTS]
  version = 3
  auth_password = "TOKEN"
  auth_protocol = "MD5"
  priv_password = "TOKEN"
  priv_protocol = "AES"
  sec_name = "test"
  sec_level = "authPriv"
  max_repetitions = 10
  timeout = "10s"
  retries = 1
  name_prefix = "dev."

  [[inputs.snmp.field]]
    name = "hostname"
    oid = ".1.3.6.1.2.1.1.5.0"
    is_tag = true
  [[inputs.snmp.field]]
    name = "sysUpTimeInstance"
    oid = "1.3.6.1.2.1.1.3.0"

    [[inputs.snmp.field]]
      name = "powerStatus"
      oid = "SNMPv2-SMI::enterprises.9.9.305.1.1.1.0"

    ### SNMPv2-SMI Section Is Not Working###
    [[inputs.snmp.field]]
      name = "csesyscpuutilization"
      oid = "1.3.6.1.4.1.9.9.305.1.1.1"
    [[inputs.snmp.field]]
      name = "cseSysUpTime"
      oid = "1.3.6.1.4.1.9.9.305.1.1.10"
    [[inputs.snmp.field]]
      name = "cseSysMemoryUtilization"
      oid = "1.3.6.1.4.1.9.9.305.1.1.2"
    [[inputs.snmp.field]]
      name = "PowerSupplyStatus"
      oid = "1.3.6.1.4.1.9.9.117.1.1.2"
    [[inputs.snmp.field]]
      name = "PowerSupplyValues"
      oid = "1.3.6.1.4.1.9.9.117.1.1.4"
    #########################################

##Everything below sends metrics correctly to Wavefront##
[[inputs.snmp.field]]
    name = "hostname"
    oid = ".1.3.6.1.2.1.1.5.0"
    is_tag = true
  [inputs.snmp.tagdrop]
    ifName=["Vlan[0-9]*", "loopback[0-9]"]
    fieldpass = ["ifInUcastPkts", "ifInNUcastPkts", "ifInDiscards", "ifInErrors", "ifOutUcastPkts", "ifOutNUcastPkts", "ifOutDiscards", "ifOutErrors", "ifType", "ifMtu", "ifSpeed", "ifAdminStatus", "ifOperStatus"]
  [[inputs.snmp.table]]
    name = "interfaceStats"
    inherit_tags = ["hostname"]
  [inputs.snmp.tagdrop]
    ifAlias = [""]
    [[inputs.snmp.table.field]]
      name = "ifAlias"
      oid = ".1.3.6.1.2.1.31.1.1.1.18"
      is_tag = true
    [[inputs.snmp.table.field]]
      name = "ifName"
      oid = ".1.3.6.1.2.1.31.1.1.1.1"
      is_tag = true
    [[inputs.snmp.table.field]]
	    name = "ifType"
  	  oid = ".1.3.6.1.2.1.2.2.1.3"
    [[inputs.snmp.table.field]]
      name = "ifInNUcastPkts"
      oid = "1.3.6.1.2.1.2.2.1.12"
    [[inputs.snmp.table.field]]
      name = "ifInDiscards"
      oid = "1.3.6.1.2.1.2.2.1.13"
    [[inputs.snmp.table.field]]
      name = "ifInErrors"
      oid = "1.3.6.1.2.1.2.2.1.14"
    [[inputs.snmp.table.field]]
      name = "ifOutUcastPkts"
      oid = "1.3.6.1.2.1.2.2.1.14"
    [[inputs.snmp.table.field]]
      name = "ifOutNUcastPkts"
      oid = "1.3.6.1.2.1.2.2.1.18"
    [[inputs.snmp.table.field]]
      name = "ifOutDiscards"
      oid = "1.3.6.1.2.1.2.2.1.19"
    [[inputs.snmp.table.field]]
      name = "ifOutErrors"
      oid = "1.3.6.1.2.1.2.2.1.20"
    [[inputs.snmp.table.field]]
      name = "ifMtu"
      oid = "1.3.6.1.2.1.2.2.1.4"
    [[inputs.snmp.table.field]]
      name = "ifSpeed"
      oid = "1.3.6.1.2.1.2.2.1.5"
    [[inputs.snmp.table.field]]
      name = "ifAdminStatus"
      oid = "1.3.6.1.2.1.2.2.1.7"
    [[inputs.snmp.table.field]]
      name = "ifOperStatus"
      oid = "1.3.6.1.2.1.2.2.1.8"
```

**Comment by Gelob** [comment]
@js-mode This likely isn't an issue with telegraf. 

TLDR; if you add .0 to the end of your OIDs for 1.3.6.1.4.1.9.9.305 that should fix your issue. Think of the .0 as an index. I imagine it might be the same for PowerSupplyStatus and PowerSupplyValues but i didn't see you include an snmpwalk output from 1.3.6.1.4.1.9.9.117. If you have more than 1 power supply and you get more than one result when you do an snmpwalk and you don't want to hardcode each one you should be able to do something like below:

```
[[inputs.snmp.table.field]]
  name = "PowerSupplyStatus"
  oid = "1.3.6.1.4.1.9.9.117.1.1.2.1.2"
```
```
    ### SNMPv2-SMI Section Is Not Working###
    [[inputs.snmp.field]]
      name = "csesyscpuutilization"
      oid = "1.3.6.1.4.1.9.9.305.1.1.1.0"
    [[inputs.snmp.field]]
      name = "cseSysUpTime"
      oid = "1.3.6.1.4.1.9.9.305.1.1.10.0"
    [[inputs.snmp.field]]
      name = "cseSysMemoryUtilization"
      oid = "1.3.6.1.4.1.9.9.305.1.1.2.0"
    #########################################
```


You shouldn't worry about snmptranslate unless you are specifying the MIB name in your oid = entry. Since you are using the OID any issue with snmptranslate isn't relevant.

snmpget is very specific and will only return results from the specific OID you supplied, where as snmpwalk will "walk" everything underneath the OID you specify. I always find it easier to do an snmpwalk when troubleshooting so that I understand all of the data I can work with and then get the specific OID.

By default snmpwalk will convert the OID to MIB, thats why when you get your results you see SNMPv2-SMI::enterprises instead of 1.3.6.1.4.1. You can turn that off in your snmpwalk by using the options -On.  `snmpwalk -On -v2c -c simple IP 1.3.6.1.4.1.9.9.305`

You get the error "No Such Instance currently exists at this OID" when you run `snmpget -v2c -c simple _IP_ 1.3.6.1.4.1.9.9.305.1.1.2` because there is no data being returned from the device with that OID. Thats normal if the device doesn't support the MIB or you have an incorrect OID. When you ran your snmpwalk against the entire ciscoSystemExtMIB tree `snmpwalk -v2c -c simple IP 1.3.6.1.4.1.9.9.305` you got back a result like this: SNMPv2-SMI::enterprises.9.9.305.1.1.2.0 = Gauge32: 59. If you did an snmpget with 1.3.6.1.4.1.9.9.305.1.1.2.0 you shouldn't get the no such instance error.

It may help to look at the oids on oidview. You can easily see there what is a table, tree, and leaf. It may help you to compare them to the interface mib which you have working to understand the relationships.
http://www.oidview.com/mibs/9/CISCO-ENTITY-FRU-CONTROL-MIB.html
http://www.oidview.com/mibs/9/CISCO-SYSTEM-EXT-MIB.html
http://www.oidview.com/mibs/0/IF-MIB.html

**Comment by danielnelson** [comment]
@js-mode Could you give the solution from @Gelob a try and let us know how it goes?

**Comment by js-mode** [comment]
@Gelob - Thank you so much for the thorough reply.  Adding the `-On` flag to get the full OID is a huge help. I am going to close this issue because it seems to be working fine when I add the .0 (as the index), but I am still uncertain on a few points (not really telegraf related), that may be helpful to others that are just entering into the `Telegraf-SNMP` world:

In looking at the OIDview -- when I take a specific leaf and add it to an `[[inputs.snmp.field]]` within my config it still gives me this error: 
> .1.3.6.1.4.1.9.9.117.1.1.2.1.5 = No Such Object available on this agent at this OID

The "leafs" from this [page](http://www.oidview.com/mibs/9/CISCO-ENTITY-FRU-CONTROL-MIB.html) has a table:  **cefcFRUPowerSupplyGroupEntry    1.3.6.1.4.1.9.9.117.1.1.1.1**

When I run the following commands, I still get the same error: 
`snmpget -On -v 2c -c simple 10.10.0.199 .1.3.6.1.4.1.9.9.117.1.1.1.1`
>.1.3.6.1.4.1.9.9.117.1.1.1.1 = No Such Object available on this agent at this OID
`snmpget -On -v 2c -c simple 10.10.0.199 .1.3.6.1.4.1.9.9.117.1.1.1.0`
>.1.3.6.1.4.1.9.9.117.1.1.1.0 = No Such Object available on this agent at this OID
`snmpget -On -v 2c -c simple 10.10.0.199 .1.3.6.1.4.1.9.9.117.1.1.1.1`
>.1.3.6.1.4.1.9.9.117.1.1.1.1 = No Such Object available on this agent at this OID

If I run snmpwalk it will give me a bunch of metrics that are not part of the MIB:

snmpwalk -On -v 2c -c simple 10.10.0.199 .1.3.6.1.4.1.9.9.117.1.1.1.1
.1.3.6.1.4.1.9.9.117.1.1.1.1.1.116950 = INTEGER: 5
.1.3.6.1.4.1.9.9.117.1.1.1.1.2.116950 = STRING: "centiAmps @ 12V"
.1.3.6.1.4.1.9.9.117.1.1.1.1.3.116950 = INTEGER: 1500
.1.3.6.1.4.1.9.9.117.1.1.1.1.4.116950 = INTEGER: 0
.1.3.6.1.4.1.9.9.117.1.1.1.1.5.116950 = INTEGER: 5

In this response -- what does the ***.116950 referring to and how would I add this to my Telegraf config file so that it could monitor that specific OID?

Is that the `[[inputs.snmp.table.field]]` you reference in your response? If I add the full OID then it should grab all of the fields within that table? 

Super helpful on the snmptranslate.  I am not too concerned with the MIB name as much as making sure I have the proper numerical OID value.   My main question is what would be the best way to figure out if I am getting an error because I have the incorrect OID or my device doesn't support the MIB?   If there is a specific MIB that my Cisco Nexus switch is using - how do I go about making sure that I can install that on the agent Telegraf is running on (or is that even necessary)?

Sorry for the really basic questions, but just trying to get a deeper understanding of how this all works under the hood and sometimes it takes a few examples and tests for it to really sink in.

**Comment by Gelob** [comment]
@js-mode Yes with regards to inputs.snmp.table.field, if theirs X number of instances under say 1.3.6.1.4.1.9.9.117.1.1.1.1.1 it will grab all of them. Think of how your interfaceStats works. You tell telegraf ifAlias is at 1.3.6.1.2.1.31.1.1.1.18 and if you snmpwalk that you'd get a bunch of data back for every interface you have.
 
You can usually assume if you can't snmpwalk it, your device doesn't have data for that MIB. Usually what I do if I get that error is to start snmpwalking higher up in the tree. Say .1.3.6.1.4.1.9.9.117.1.1.1.1.5. doesn't return anything, ok what about .1.3.6.1.4.1.9.9.117.1.1.1.1, no? ok take off another integer. You'll obviously be getting different data but you can understand if you are just missing something or if that MIB isn't supported by your device pretty easily.

I couldn't seem to figure out what 116950 references even digging into the MIB.

**Comment by js-mode** [comment]
Thank you for following up even though the issue has been closed.  I now have a Telegraf question/issue which is still causing me trouble that I would appreciate help with. 

It seems as though if I add this -- I get an error from Telegraf saying:

>2019-04-02T19:26:00Z E! [inputs.snmp]: Error in plugin: initializing table PowerSupply: getting table columns: exit status 1: Was that a table? SNMPv2-SMI::enterprises.9.9.117.1.1

What I added to my config file was:
``` 
[[inputs.snmp.table]]
   name = "PowerSupply"
   oid = ".1.3.6.1.4.1.9.9.117.1.1"
```

When I add it as [[inputs.snmp.field]] instead, nothing happens at all. I don't get any input or error from Telegraf. 

I can run an snmpwalk against oid = ".1.3.6.1.4.1.9.9.117.1.1" and I get the following response:
```
.1.3.6.1.4.1.9.9.117.1.1.1.1.1.116950 = INTEGER: 5
.1.3.6.1.4.1.9.9.117.1.1.1.1.2.116950 = STRING: "centiAmps @ 12V"
.1.3.6.1.4.1.9.9.117.1.1.1.1.3.116950 = INTEGER: 5416
.1.3.6.1.4.1.9.9.117.1.1.1.1.4.116950 = INTEGER: 1652
.1.3.6.1.4.1.9.9.117.1.1.1.1.5.116950 = INTEGER: 5
.1.3.6.1.4.1.9.9.117.1.1.2.1.1.470 = INTEGER: 1
.1.3.6.1.4.1.9.9.117.1.1.2.1.1.471 = INTEGER: 1
.1.3.6.1.4.1.9.9.117.1.1.2.1.2.470 = INTEGER: 2
.1.3.6.1.4.1.9.9.117.1.1.2.1.2.471 = INTEGER: 2
.1.3.6.1.4.1.9.9.117.1.1.2.1.3.470 = INTEGER: 5416
.1.3.6.1.4.1.9.9.117.1.1.2.1.3.471 = INTEGER: 5416
```

So my question now is how would I add this OID (which is a [valid OID](http://www.oidview.com/mibs/9/CISCO-ENTITY-FRU-CONTROL-MIB.html) that returns information) to my existing config file? 

Could it be that I need to have a separate [[inputs.snmp]] for each of the MIB/OID tables I am pulling from on my switch?   The Interface MIBs all have the prefix of ".1.3.6.1.2.1.31.***", the CSE MIBs all have the prefix of ".1.3.6.1.4.1.9.9.305.***" and the FRU Power MIBs all have the prefix of ".1.3.6.1.4.1.9.9.117.1.***" 

So in this instance, is it required to have separate [[inputs.snmp]] for each? I am not sure how this will work with the time out and the `agent intervals` - but I feel like there should be a way to have 3 separate MIBs all polled within the same [[inputs.smp]] section of the config file? 

I realize that the order in which the tables and fields get added are important, but I am unable to figure out why can't get all 3 of them working together. 

Here is a link to the latest config file that I am using: https://gist.github.com/js-mode/0668804aedc573c0d0aa803f483edb2f

Any suggestions would really be appreciated!

**Comment by danielnelson** [comment]
Based on your link though, I think the table is `.1.3.6.1.4.1.9.9.117.1.1.1`.

The `snmptable` command is a closer analog to how the tables work in Telegraf:
```
snmptable -Cl -v2c -c public 127.0.0.1:161 .1.3.6.1.4.1.9.9.117.1.1.1
```

In order for this to work, you need the MIB available to the net-snmp tools: `snmptable` and `snmptranslate`.  You won't need a separate `[[inputs.snmp]]`.

This page does a good job of describing how Tables are represented in the MIBs tree structure:
https://intronetworks.cs.luc.edu/current/html/netmgmt.html#snmp-tables

**Comment by js-mode** [comment]
If I run this command I still get the "Was that a `table?` error"...

2019-04-02T22:55:08Z I! [agent] Config: Interval:1m0s, Quiet:false, Hostname:"telegraf-snmp", Flush Interval:10s
2019-04-02T22:56:00Z E! [inputs.snmp]: Error in plugin: initializing table PowerSupply: getting table columns: exit status 1: Was that a table? SNMPv2-SMI::enterprises.9.9.117.1.1.1
2019-04-02T22:56:59Z E! [inputs.snmp]: Error in plugin: initializing table PowerSupply: getting table columns: exit status 1: Was that a table? SNMPv2-SMI::enterprises.9.9.117.1.1.1

And if I run the `snmptable -Cl -v2c -c simple 10.10.0.199 .1.3.6.1.4.1.9.9.117.1.1.1` command I also get the "Was that a table?"

I have tried it with "*.117.1", "*.117.0" all the way to "*.117.1.1.1.1" and "*.117.1.1.1.0" using this command and every possible variance in between and I get the same "Was that a table?"

So I guess at this point I need to focus on getting the MIB loaded into the Telegraf agent.  How do I go about doing this?  I have net-snmp tools installed and all of the other Cisco MIBs seem to be working fine, but I am only having issues with with the [FRU-CONTROL-MIB](http://www.oidview.com/mibs/9/CISCO-ENTITY-FRU-CONTROL-MIB.html) for some reason? 

@Gelob  - any suggestions on how to add these two OIDs to my Telegraf.conf file so I can monitor these like the rest? 

 !!!!!! CISCO-ENTITY-FRU-CONTROL-MIB !!!!!!
Power Supply Status: 1.3.6.1.4.1.9.9.117.1.1.2
Power Supply Values: 1.3.6.1.4.1.9.9.117.1.1.4

**Comment by danielnelson** [comment]
The easiest way to install MIB files on most systems is to place them into the `/usr/share/snmp/mibs`, make sure to set the correct permissions.

May be useful to compare the output against these standard tables:
```
$ snmptable -Cl -v2c -c public 127.0.0.1:161 IF-MIB::ifTable
```

Translate should also work:
```
$ snmptranslate -Td -Ob -m all .1.3.6.1.2.1.2.2
IF-MIB::ifTable
ifTable OBJECT-TYPE
  -- FROM       IF-MIB, RFC1213-MIB
  MAX-ACCESS    not-accessible
  STATUS        current
  DESCRIPTION   "A list of interface entries.  The number of entries is
            given by the value of ifNumber."
::= { iso(1) org(3) dod(6) internet(1) mgmt(2) mib-2(1) interfaces(2) 2 }
```

**Comment by js-mode** [comment]
Ok thank you so much for getting back to me. It turns out that I couldn't get anything to show up because I forgot to add these additional fields to the "fieldpass" section... Doh 🤦‍♂️ 

So as a general note, if you are trying to get OIDs to work, comment out the fieldpass line momentarily until you get it right and then you can add it back in afterwards. 

This wasn't an issue before because the fieldpass was within the snmp.table section (incorrectly) so seems like it was just letting everything in.  But then when I moved it up in the config file, that is when it started doing its gatekeeping job... 

Thank you for all of your help @danielnelson and @Gelob  - I am hopeful that maybe someone else can learn from my rookie mistakes.
